### ISSUE TITLE
Building the model (via `gms::singleGAMSfile`) is too slow

### ISSUE THREAD

**Issue by mikapfl** [remindmodel/remind#1071]
Moving the discussion from https://github.com/remindmodel/remind/pull/1063:

## tl;dr
Building + running the model normally via `start.R --quick` is much too slow. For a simple syntax error (missing semicolon in `main.gms`), it takes 3:10 minutes on my laptop, of which 2:30 min are spent in `gms::singleGAMSfile()`. Running `gams main.gms -a=c` straight only takes 8 seconds and finds the same syntax error. For development, it would be great if simple syntax errors could be found quickly, and in general, it would be great if building the model was significantly faster.

## Discussion moved from the PR:
I wrote:
> I have a bigger question: compiling directly on main.gms (or equivalent) as done here can be quite useful and fast, but unfortunately, it is kind of brittle because it is such a different mode of operation compared to compiling full.gms. Essentially, we are writing code for two compilers: `gms::singleGAMSfile`, and `gams`. And e.g. `start.R -q` only tests if the code works when compiled with `gms::singleGAMSfile`.

@orichters wrote:
> Well, we use the tool `gamscompile` on the cluster which does exactly that… So I don't think the fact that we use both tools is new, but maybe new in this repo.

I wrote:
> No, it is not new at all (is even documented in the tutorials), it is just getting more visible. That's why I think we can merge now, but I also think longer-term we need to do something else (making `gms::singleGAMSfile` faster, abandoning full.gms, something which doesn't involve this dual-strategy).

@0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q wrote:
> > Essentially, we are writing code for two compilers: `gms::singleGAMSfile`, and `gams`. And e.g. `start.R -q` only tests if the code works when compiled with `gms::singleGAMSfile`.
> 
> `gms::singleGAMSfile()` is not a compiler but the current Guiness Book of World Records holder for "slowest preprocessor ever conceived by sentient beings within the local group".  And the original _raison d'être_ for `gamscompile`.
> 
> > I have a bigger question: compiling directly on `main.gms` (or equivalent) as done here can be quite useful and fast, but unfortunately, it is kind of brittle because it is such a different mode of operation compared to compiling `full.gms`. 
> 
> That is a statement, not a question.  And `gamscompile` is not brittle, since it simply calls `gams -a=c`, which is the ultimate arbiter on whether GAMS code compiles or not.  Whould `gms::singleGAMSfile()` diverge and do something different, the code will not run anyway.
> 
> > I also think longer-term we need to do something else (making singleGAMSfile faster, abandoning `full.gms`, something which doesn't involve this dual-strategy).
> 
> What's the dual strategy here?

I wrote:
> > That is a statement, not a question.
> 
> True, sorry.
> 
> > `gms::singleGAMSfile()` is not a compiler but the current Guiness Book of World Records holder for "slowest preprocessor ever conceived by sentient beings within the local group". And the original raison d'être for `gamscompile`
> 
> Yes, that is the outside-visible problem. Call it a preprocessor or a compiler, what I mean is: it is something which is between GAMS source code as written by humans and GAMS code as executed during GAMS runtime. Both the GAMS compilation phase and `gms::singleGAMSfile()` are in that space. Both (try to) resolve `include`s and do similar code transformation. Also, other parts of `gms` are in that space, e.g. `manipulateConfig`.
> 
> > What's the dual strategy here?
> 
> The dual strategy is to have:
> 1. the `gms::singleGAMSfile()` strategy, which is (currently, anyway) mandatory for production runs, slow, and hard to debug. It yields an actually solvable model in the output folder and a `full.gms` as artefact for future use.
> 2. the `gams -a=c` strategy, which is fast, but does not yield an actually solvable model, e.g. because it doesn't find R scripts etc. Also, it needs some setup by a previous "normal" run (input data distribution, mainly).
> 
> In an ideal world, our default "production" strategy would be as fast as `gamscompile`. I think it is technically possible to have our cake and eat it too, here. We could produce a faster preprocessor (loosing our Guinness Book of World Records entry), or not preprocess the GAMS source at all before handing it to gams, or something. We are not the only people in the world who need the functionality offered by `gms::singleGAMSfile()`, and maybe we can just use the [save and restart](https://www.gams.com/latest/docs/UG_SaveRestart.html) functions of GAMS. I haven't looked closely into the use cases of `full.gms` and if they would be covered by the save and restart feature, but I'm pretty sure we could be as fast as `gamscompile` in our default production runs.
> 
> Sorry for derailing this pull request discussion, we probably should move this to an issue or so.

@0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q wrote:
> > Sorry for derailing this pull request discussion, we probably should move this to an issue or so.
> 
> Until you do …
> 
> > 2. the `gams -a=c` strategy, which is fast, but does not yield an actually solvable model, e.g. because it doesn't find R scripts etc. Also, it needs some setup by a previous "normal" run (input data distribution, mainly).
> 
> It's not intended to yield a solvable model.  It's just intended to check for GAMS compilation errors.  Waiting more than three minutes for simple syntax checking instead of ten seconds is just ridiculous.

**Comment by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q** [comment]
Lots of "he said, he said, he said"

But shouldn't this be in the [pik-paim/gms](https://github.com/pik-piam/gms) repository?

**Comment by mikapfl** [comment]
https://github.com/pik-piam/gms/pull/47

**Comment by mikapfl** [comment]
We now have a faster implementation of compilation, and also `gamscompile` for testing purposes.
